doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,314 @@
+
+We want to make it easier for folks to get data into Materialize. Currently the only way is via one
+of our [supported sources](https://materialize.com/docs/sql/create-source/), which depending on
+the environment, can be pretty hard to setup. A common method of publishing data from an
+application is via [webhooks](https://www.redhat.com/en/topics/automation/what-is-a-webhook). When
+some event occurrs, e.g. a new analytics event from Segment, or a new Pull Request on a GitHub
+repository, an HTTP endpoint you define will get called so you can receive the event in near
+real-time.
+
+Supporting webhooks in Materialize allows users to easily get their data into MZ, and the low
+latency of the events works very well with the premise of being an "operational" data warehouse.
+
+> **_Note_** To achieve the goal of "make it easier to get data in Materialize" we also considered
+building support for Fivetran. We may still do this in the future, but the amount of work to get
+webhooks working is significantly less, and the frequency of data is more suited to the goals of
+Materialize.
+
+## Goals
+
+1. Provide URLs that users can push data to.
+2. Support an extensible syntax similar to `CREATE SOURCE <name> FROM WEBHOOK` that allows
+users to generate the aforementioned URLs.
+3. Handle at least a total of 100 requests per-second across all `WEBHOOK` sources.
+5. Code complete by August 15th, 2023.
+
+
+## Non-Goals
+
+The following are all features we want to support, but are out of scope in the immediate term.
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+* Handling batch requests.
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `WEBHOOK` source, and add a new endpoint to
+`environmentd`'s HTTP server `/api/webhook/:database/:schema/:name`. When a user creates this new
+kind of source, we'll persist the new `Source` in the catalog, use the `storage-client` to create a
+new collection.
+
+When we receive a event we'll use the `mz_adapter::Client` that the HTTP server has to send the raw
+`headers` and `body` to the coordinator, pack the data into a row, and append it to the necessary
+collection using a new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `WEBHOOK`, for which you can
+specify a `BODY FORMAT`, this denotes how we decode the body of a request. To start we'll support
+three formats, `BYTES`, `JSON`, or `TEXT`. By default we will not include the headers of the
+request, they can optionally be added by specifying `INCLUDE HEADERS`.
+
+```
+CREATE SOURCE <name> FROM WEBHOOK
+  BODY FORMAT [BYTES | JSON | TEXT]
+  (INCLUDE HEADERS)?
```

Comment:
This is nbd, but I might suggest using the existing style `FooOption` structs that would make this look more like:

```
CREATE SOURCE <name> FROM WEBHOOK
  BODY FORMAT [BYTES | JSON | TEXT]
  (WITH (HEADERS))?
```

This way if we add any options in the future we're not asking folks to remember which things are options and which of them have their own dedicated bits of syntax. For an example of this, you can look up `PgConfigOption`.

If there's been discusion about this I missed, this is isn't blcoking.

Reply:
Looking over the rest of the proposal, seems like headers might be more integral to the source than an option suggestions + might be awkward with the rest of the proposed syntax that might want to use the header column. The above is probably not the right tack.

Reply:
We just messed this up in the Kafka syntax, is my take. `KEY FORMAT`, `VALUE FORMAT`, `INCLUDE HEADERS`, `ENVELOPE` should all be options that get attached to the specific source type rather than top-level options. But they are top-level source options rather than Kafka-specific options.

Context: https://github.com/MaterializeInc/materialize/pull/20002#discussion_r1237985877

I'm torn. If there is broad agreement we messed this up for Kafka, we could instead choose to do the better thing for webhook sources. But in that case I'd want to also improve the Kafka source syntax to support putting the options inside parenthesis:

```
CREATE SOURCE FROM KAFKA (
    KEY FORMAT ...
    VALUE FORMAT ...
    INCLUDE HEADERS ...
)
```

Reply:
I am in the camp that these things should be options rather than top-level keywords. Not a hill I'd be willing to die on but code I'd be willing to write.

Reply:
I'm very on board with trying to fix this, Sean! But I think the next step is a design doc that proposes a holistic syntax improvement for all two (three) of our sources. E.g. where does `ENVELOPE` go?  What do we do about the fact that `INCLUDE HEADERS [AS ALIAS]` is a thing? Most `CREATE SOURCE` options don't neatly map to our simple "with options" framework.

Reply:
The limiting factor in the options framework is `WithOptionValue`, which can be extended to contain whatever we need.

For example, we could have something like:

```rust
pub enum WithOptionValue<T: AstInfo, P: ParseableEnum> {
   ...
   ParseableEnum(Vec<Keyword>),
}
```

Ee could then post-facto take those keywords into the enums we want. We can't do something more reasonable and make this a trait because of the visitor generator, though I'm sure with enough elbow grease that could be mitigated, as well.

Sub options are more gross--if `HEADER ALIAS` require `INCLUDE HEADERS`?--but there's a push and pull here: we could make this a little more extensible/clean, but as a framework it will require some different compromises.

I can work on a design doc for this next week.

Reply:
> We can't do something more reasonable and make this a trait because of the visitor generator, though I'm sure with enough elbow grease that could be mitigated, as well.

That sounds totally fine to me, FWIW. I don't mind the code being complicated for the complicated cases! I think the hard questions are mostly on the syntax design side. Do we keep the existing syntax of `INCLUDE HEADER [AS alias]`, and just move it inside the parentheses, or do we split it into the two options you mention (`INCLUDE HEADER` and `HEADER ALIAS`)?

> I can work on a design doc for this next week.

🙇🏽 

Say, while you're in the area, it would be great to tackle the following issues too:

  * #15272
  * #19143 (just the part about redesigning the syntax)

Reply:
I filed a tracking issue to get my thoughts down: https://github.com/MaterializeInc/materialize/issues/20135

Reply:
@benesch @sploiselle how do y'all feel about proceeding with the proposed syntax, with the entire feature behind a LaunchDarkly flag? This way Sean can proceed with the design doc in parallel to this implementation, and before we release it to any users we can sync again on the syntax?

Reply:
Exactly my thoughts--this shouldn't be blocking until GA.

Reply:
Sounds great!

Reply:
Opened https://github.com/MaterializeInc/materialize/issues/20217 to make sure we circle back on this before releasing webhook sources to GA